### ISSUE TITLE
Should Node (and its sub-classes) take children as argument to constructor

### ISSUE THREAD

**Issue by arporter** [stfc/PSyclone#645]
`Node.__init__` takes a list of children as one of its arguments and stores them under `self._children`. However, it makes no attempt to set the parent of those child nodes.

In contrast, `Loop.__init__` and `IfBlock.__init__` do *not* have `children` as an argument to their constructors. They do have static `create` methods that, given suitable children, construct a fully-populated Loop or IfBlock object.

We therefore need to agree whether or not a Node (and its subclasses) should have `children` as an argument to the constructor and if so, whether or not the constructor should also set the parent of those nodes. Searching for 'children=' in psyGen.py, it seems that the only real use of the `children` constructor argument is when making `Schedule`s, i.e. the "PSy-layer" PSyIR.

One of the main ways we construct PSyIR is through a visitor that walks down the parse tree. We therefore are often creating the parent node before we proceed to create any children. Since the visitor updates the PSyIR tree in place (rather than returning e.g. a list of child nodes) this parent has to exist before we can proceed with the walk down the tree.
        
It therefore seems likely that we should drop the `children` argument from the constructor and provide other ways of setting things up (that can also help to ensure internal consistency in parent-child relationships).

**Comment by arporter** [comment]
Note that some use of the `children` constructor argument is also made in transformations.py. However, this is not extensive (~5-6 instances) and is confined to the creation of Directives.

**Comment by hiker** [comment]
I totally agree to have ``set_parent`` and ``add_child`` functions, which will also clean up existing parent/child relations. The ``add_child`` function should take either a single node, or a list of nodes I'd say (though ... should it then be named ``add_children``? Or two different functions, add_children just repeatedly calling add_child?)

I am neutral if we should have a parent/child optional parameter in the constructors of all node types. It is convenient (I am thinking of transformations, which often need to insert a node to an existing hierarchy - thought we might as well add a special ``insert_node`` function for that), but a lot of work to add these parameters to all nodes, so from my point of view I am happy to consistently remove them (or add them) - but whatever we do, it should set up/clean up all parent/child relations!

**Comment by arporter** [comment]
I've come across this issue again while reviewing @hiker's work on the PSyData node. When creating Extract/Data/Profile nodes we rely on passing both parent and children to the constructor and it then sets-up the necessary links. This is slightly non-intuitive (for me at least). However, this particular situation is complicated by the fact that when creating an e.g. ExtractNode, we have to create a Schedule as its first child and then give that Schedule all of the provided 'child' nodes.
